File: testdata/emitted-packages/11-cohen-h3k27ac-crc/runtime/plotting_r/stages/quality_control.R
# R-side quality_control stage figures: per_sample_metric_violin,
# per_sample_metric_bar, qc_summary_bar.
# Mirrors lib/plotting/stages/quality_control.py contract.
#
# Plan reference: Phase 10 affordance catalog expansion;
# semantic_type data:2914 → renderer_module
# runtime.plotting.stages.quality_control.
#
# Expected inputs (any subset works; missing inputs produce a typed
# error recorded by ecaa_generate, not a crash):
#
#   outputs_dir/qc_metrics.tsv[.gz]  — long-form: sample, metric, value
#   manifest.per_sample_metrics      — {sample_id: {metric: scalar}}
#   outputs_dir/summary_stats.json   — {key: numeric_value, ...}

if (!exists("ecaa_register_figure")) {
  stop("source runtime/plotting_r/core.R before this stage module")
}

# ---------------------------------------------------------------------------
# Internal data loaders — mirrors Python _load_long_metrics /
# _manifest_per_sample helpers.
# ---------------------------------------------------------------------------

# Parse qc_metrics.{tsv,tsv.gz} → list(metric = list(sample = values)).
# Returns NULL when the file is missing or the header is unrecognised.
.qc_load_long_metrics <- function(outputs_dir) {
  for (nm in c("qc_metrics.tsv.gz", "qc_metrics.tsv")) {
    p <- file.path(outputs_dir, nm)
    if (!file.exists(p)) next
    df <- tryCatch(
      utils::read.delim(p, stringsAsFactors = FALSE, check.names = FALSE,
                        comment.char = ""),
      error = function(e) NULL
    )
    if (is.null(df) || nrow(df) == 0L) next
    cols_lc <- tolower(colnames(df))
    i_sample <- which(cols_lc == "sample")
    i_metric <- which(cols_lc == "metric")
    i_value  <- which(cols_lc == "value")
    if (!length(i_sample) || !length(i_metric) || !length(i_value)) return(NULL)
    out <- list()
    for (row_i in seq_len(nrow(df))) {
      val <- suppressWarnings(as.numeric(df[[i_value[1]]][row_i]))
      if (is.na(val)) next
      metric <- as.character(df[[i_metric[1]]][row_i])
      sample <- as.character(df[[i_sample[1]]][row_i])
      if (!is.list(out[[metric]])) out[[metric]] <- list()
      if (!is.numeric(out[[metric]][[sample]])) {
        out[[metric]][[sample]] <- val
      } else {
        out[[metric]][[sample]] <- c(out[[metric]][[sample]], val)
      }
    }
    return(if (length(out) > 0L) out else NULL)
  }
  NULL
}

# Pull per-sample scalar metrics out of manifest.per_sample_metrics.
# Returns list(samples = character, metrics = list(name = numeric_vec))
# or NULL when the manifest block is absent.
.qc_manifest_per_sample <- function(ctx) {
  per_sample <- ctx$manifest$per_sample_metrics
  if (!is.list(per_sample) || length(per_sample) == 0L) return(NULL)
  samples <- sort(names(per_sample))
  # Collect all numeric metric names across samples.
  metric_names <- character(0)
  for (s in samples) {
    m <- per_sample[[s]]
    if (is.list(m)) {
      metric_names <- union(metric_names,
        names(Filter(function(v) is.numeric(v) || is.integer(v), m)))
    }
  }
  if (length(metric_names) == 0L) return(NULL)
  metric_names <- sort(metric_names)
  metrics <- lapply(stats::setNames(metric_names, metric_names), function(mn) {
    sapply(samples, function(s) {
      v <- per_sample[[s]][[mn]]
      if (!is.null(v) && (is.numeric(v) || is.integer(v))) as.numeric(v)
      else NA_real_
    })
  })
  list(samples = samples, metrics = metrics)
}

# ---------------------------------------------------------------------------
# per_sample_metric_violin
# Violin (with embedded boxplot) of per-sample QC metrics across samples.
# Falls back to a grouped bar when only scalar per-sample data is available
# (no long-form qc_metrics.tsv distribution), matching the Python fallback.
# ---------------------------------------------------------------------------

ecaa_register_figure("quality_control", "per_sample_metric_violin",
                     function(ctx) {
  long_form <- .qc_load_long_metrics(ctx$outputs_dir)
  if (!is.null(long_form) && length(long_form) > 0L) {
    # Pick the metric with the most distinct values (widest variation).
    metric_name <- names(long_form)[which.max(
      sapply(long_form, function(m) {
        all_vals <- unlist(m)
        if (is.null(all_vals)) return(0L)
        length(unique(all_vals))
      })
    )]
    samples_per_metric <- long_form[[metric_name]]
    data_list <- lapply(samples_per_metric, as.numeric)
    p <- ecaa_violin(
      data_list,
      title   = sprintf("QC: %s per sample", metric_name),
      ylabel  = metric_name,
      x_label = "sample"
    )
    ecaa_savefig(p, path = file.path(ctx$outputs_dir, "figures",
                                      "per_sample_metric_violin.png"),
                 stage_id = "quality_control")
    return(p)
  }
  # Fallback: scalar manifest data → bar.
  mdata <- .qc_manifest_per_sample(ctx)
  if (is.null(mdata)) {
    stop("no qc_metrics.tsv[.gz] or manifest.per_sample_metrics")
  }
  metric_name <- names(mdata$metrics)[1]
  p <- ecaa_bar(
    names  = mdata$samples,
    values = mdata$metrics[[metric_name]],
    title  = sprintf("QC: %s per sample", metric_name),
    ylabel = metric_name,
    xlabel = "sample"
  )
  ecaa_savefig(p, path = file.path(ctx$outputs_dir, "figures",
                                    "per_sample_metric_violin.png"),
               stage_id = "quality_control")
  p
})

# ---------------------------------------------------------------------------
# per_sample_metric_bar
# Horizontal bar of the cardinal per-sample count metric. Prefers a key
# containing the substring "count" or starting with "n_"; falls back to
# the first key. Matches Python per_sample_metric_bar.
# ---------------------------------------------------------------------------

ecaa_register_figure("quality_control", "per_sample_metric_bar",
                     function(ctx) {
  mdata <- .qc_manifest_per_sample(ctx)
  if (is.null(mdata)) {
    stop("manifest.per_sample_metrics required for per_sample_metric_bar")
  }
  metric_keys <- names(mdata$metrics)
  preferred <- {
    cnt_key <- grep("count",           metric_keys, value = TRUE,
                    ignore.case = TRUE)[1]
    n_key   <- grep("^n_",             metric_keys, value = TRUE,
                    ignore.case = TRUE, perl = TRUE)[1]
    if (!is.na(cnt_key)) cnt_key
    else if (!is.na(n_key)) n_key
    else metric_keys[1]
  }
  p <- ecaa_bar(
    names      = mdata$samples,
    values     = mdata$metrics[[preferred]],
    title      = sprintf("QC: %s per sample", preferred),
    ylabel     = preferred,
    xlabel     = "sample",
    horizontal = TRUE          # horizontal bar for per-sample count
  )
  ecaa_savefig(p, path = file.path(ctx$outputs_dir, "figures",
                                    "per_sample_metric_bar.png"),
               stage_id = "quality_control")
  p
})

# ---------------------------------------------------------------------------
# qc_summary_bar
# Aggregate summary bar from summary_stats.json. Useful when the stage
# didn't emit per-sample distributions but produced aggregate counts
# (e.g. total reads, total variants, total cells). Matches Python
# qc_summary_bar.
# ---------------------------------------------------------------------------

ecaa_register_figure("quality_control", "qc_summary_bar", function(ctx) {
  summary_path <- file.path(ctx$outputs_dir, "summary_stats.json")
  if (!file.exists(summary_path)) stop("summary_stats.json")
  summary_data <- tryCatch(
    jsonlite::fromJSON(summary_path, simplifyVector = TRUE),
    error = function(e) stop(sprintf("summary_stats.json unparseable: %s",
                                      conditionMessage(e)))
  )
  # Keep only scalar (length-1) numerics.
  scalar <- Filter(
    function(v) is.numeric(v) && length(v) == 1L && is.finite(v),
    summary_data
  )
  if (length(scalar) == 0L) stop("summary_stats.json has no scalar metrics")
  nm  <- names(scalar)
  val <- as.numeric(unlist(scalar))
  p <- ecaa_bar(
    names  = nm,
    values = val,
    title  = "QC: aggregate summary",
    ylabel = "value"
  )
  ecaa_savefig(p, path = file.path(ctx$outputs_dir, "figures",
                                    "qc_summary_bar.png"),
               stage_id = "quality_control")
  p
})

File: testdata/emitted-packages/17-lloyd-price-hmp2/runtime/plotting/stages/quality_control.py
"""QC stage figures — applies across every modality that produces
per-sample quality metrics (bulk RNA-seq, scRNA-seq, ChIP-seq, variant
calling, proteomics, metagenomics, long-read RNA-seq).

Expected inputs (in outputs_dir, any subset works):
- manifest.json with a "per_sample_metrics" map of sample_id → metric dict
- qc_metrics.tsv / qc_metrics.tsv.gz: long-form rows with columns
  `sample`, `metric`, `value`
- summary_stats.json with aggregate numbers

Stage is tolerant: missing inputs → the figure lands in `skipped`, not
`errors`, so the stage still produces every figure it can.
"""

from __future__ import annotations

import gzip
import json
from pathlib import Path
from typing import Dict, List, Optional, Tuple

import numpy as np

from ..core import FigureContext, bar, stage_registry, violin, register_figure

FIGURES = stage_registry("quality_control")


def _load_long_metrics(
    outputs_dir: Path,
) -> Optional[Dict[str, Dict[str, List[float]]]]:
    """Parse qc_metrics.{tsv,tsv.gz} → {metric: {sample: [values]}}. Returns
    None if the file is missing or unparseable so the caller can fall
    through to the manifest-derived path.
    """
    for name in ("qc_metrics.tsv.gz", "qc_metrics.tsv"):
        p = outputs_dir / name
        if not p.exists():
            continue
        opener = gzip.open if name.endswith(".gz") else open
        try:
            with opener(p, "rt") as f:
                header = f.readline().strip().split("\t")
                try:
                    i_sample = header.index("sample")
                    i_metric = header.index("metric")
                    i_value = header.index("value")
                except ValueError:
                    return None
                out: Dict[str, Dict[str, List[float]]] = {}
                for line in f:
                    parts = line.rstrip("\n").split("\t")
                    if len(parts) <= max(i_sample, i_metric, i_value):
                        continue
                    try:
                        val = float(parts[i_value])
                    except ValueError:
                        continue
                    metric = parts[i_metric]
                    sample = parts[i_sample]
                    out.setdefault(metric, {}).setdefault(sample, []).append(val)
                return out
        except OSError:
            return None
    return None


def _manifest_per_sample(ctx: FigureContext) -> Optional[Tuple[List[str], Dict[str, List[float]]]]:
    """Pull per-sample scalar metrics out of manifest.json. Returns
    (samples, {metric: values-in-sample-order}) or None.
    """
    per_sample = ctx.manifest.get("per_sample_metrics")
    if not isinstance(per_sample, dict) or not per_sample:
        return None
    samples = sorted(per_sample.keys())
    # Collect metric set
    metric_names: set = set()
    for s in samples:
        if isinstance(per_sample[s], dict):
            metric_names.update(
                k for k, v in per_sample[s].items() if isinstance(v, (int, float))
            )
    out: Dict[str, List[float]] = {}
    for m in sorted(metric_names):
        vals = []
        for s in samples:
            v = per_sample.get(s, {}).get(m)
            vals.append(float(v) if isinstance(v, (int, float)) else np.nan)
        out[m] = vals
    if not out:
        return None
    return samples, out


@register_figure(FIGURES, "per_sample_metric_violin")
def per_sample_metric_violin(ctx: FigureContext, out: Path) -> Optional[Path]:
    """Violin per QC metric across samples. Falls back to a bar when we
    only have one value per sample per metric (no per-cell distribution
    in the artifact).
    """
    long_form = _load_long_metrics(ctx.outputs_dir)
    if long_form:
        # Pick the metric with the widest variation for the primary violin
        metric = next(iter(long_form))
        return violin(
            data=long_form[metric],
            title=f"QC: {metric} per sample",
            ylabel=metric,
            out=out,
            x_label="sample",
        )
    manifest_data = _manifest_per_sample(ctx)
    if manifest_data is None:
        raise FileNotFoundError("no qc_metrics.tsv[.gz] or manifest.per_sample_metrics")
    samples, metrics = manifest_data
    # Pick the first metric and render as a bar (one value per sample).
    metric_name = next(iter(metrics))
    return bar(
        names=samples,
        values=metrics[metric_name],
        title=f"QC: {metric_name} per sample",
        ylabel=metric_name,
        out=out,
        xlabel="sample",
    )


@register_figure(FIGURES, "per_sample_metric_bar")
def per_sample_metric_bar(ctx: FigureContext, out: Path) -> Optional[Path]:
    """Bar chart of the cardinal per-sample count (n_cells for scRNA,
    n_reads for bulk, n_variants for calling, etc.). Reads
    manifest.per_sample_metrics — the primary metric is whichever key
    contains the substring `count` or `n_`, else the first key.
    """
    manifest_data = _manifest_per_sample(ctx)
    if manifest_data is None:
        raise FileNotFoundError("manifest.per_sample_metrics required for per_sample_metric_bar")
    samples, metrics = manifest_data
    preferred = next(
        (k for k in metrics if ("count" in k.lower() or k.lower().startswith("n_"))),
        next(iter(metrics)),
    )
    return bar(
        names=samples,
        values=metrics[preferred],
        title=f"QC: {preferred} per sample",
        ylabel=preferred,
        out=out,
        xlabel="sample",
    )


@register_figure(FIGURES, "qc_summary_bar")
def qc_summary_bar(ctx: FigureContext, out: Path) -> Optional[Path]:
    """Bar of top-level QC summary numbers from summary_stats.json when
    present. Useful when the stage didn't emit per-sample data but
    produced aggregate counts.
    """
    summary_path = ctx.outputs_dir / "summary_stats.json"
    if not summary_path.exists():
        raise FileNotFoundError("summary_stats.json")
    try:
        summary = json.loads(summary_path.read_text())
    except json.JSONDecodeError as e:
        raise FileNotFoundError(f"summary_stats.json unparseable: {e}") from e
    scalar = {k: v for k, v in summary.items() if isinstance(v, (int, float))}
    if not scalar:
        raise FileNotFoundError("summary_stats.json has no scalar metrics")
    names = list(scalar.keys())
    values = [float(scalar[n]) for n in names]
    return bar(
        names=names,
        values=values,
        title="QC: aggregate summary",
        ylabel="value",
        out=out,
    )
